### PR TITLE
Danielsola/se 109 duck db example

### PR DESCRIPTION
**Pull request by dansola**
Copy of message sent to MotherDuck:

Here are some updates:
1. We adapted our current flyte duckdb plugin to also work with motherduck.
2. Wrote an example for a blogpost:
    - Wanted to showcase the following:
        - Efficiency of duckdb for analytical queries.
        - The ability to query in memory data and remote data in motherduck at the same time in the same query.
    - To showcase these features, I chose an ecommerce dataset that would suit very analytical queries. With some "new" incoming ecommerce data, say in a dataframe, we can also query that at the same time as remote data in motherduck to get some sales and customer trends/statists.

The current state of the example:
1. We have a Union workflow that takes in a pandas dataframe containing recent ecommerce data.
2. We run various queries that simultaneously target the local (dataframe) and remote (motherduck) data to get some information like "how are products selling compared to average?"
3. Display some plots in the Union UI showing the results of those queries.
4. There will also be an optional input into the workflow which is a prompt. This can be something simple like "how many customers bought product x?" Or something more complex. This is routed to openai function calling which prepares a duckdb query that can be ran on motherduck data and/or the local data to answer the original user prompt. This serves as a natural language to duckdb interface and follows the best practices outlined by openai.

Example Run: 
`union -c ~/.uctl/config-demo.yaml run --remote -p daniel ecommerce_wf.py wf --prompt="How many customers are there in the historical data compared to the recent data?"`
https://demo.hosted.unionai.cloud/console/projects/daniel/domains/development/executions/fc7ba109f1c754508ad4/nodeId/n4/nodes

Blog post: https://docs.google.com/document/d/1Xq3vwGyiAYUh6TlgwpgHdLGCDHGb7PdpmdF2HZo9RvE/edit

### CODE REVIEW

File: _blogs/motherduck/ecommerce_wf.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+from plots import make_elasticity_plot, make_sales_trend_plot, create_transactions_plot, \
+    create_spend_plot
+from queries import sales_trends_query, elasticity_query, customer_segmentation_query
+
+image = ImageSpec(
+    name="test-image",
+    # registry=os.environ.get("DOCKER_REGISTRY", None),
+    registry="ghcr.io/dansola",
+    apt_packages=["git"],
+    packages=["pandas==2.2.2", "plotly==5.23.0", "pyarrow==16.1.0", "openai==1.41.0", "flytekitplugins-openai"],#, "git+https://github.com/flyteorg/flytekit.git@6d5f81e81844c1a0bad7ba6614537c9afa8a4ff6#subdirectory=plugins/flytekit-duckdb"],
```
Comment (samhita-alla) [inline _blogs/motherduck/ecommerce_wf.py:23]: can you clean this up? let's also remove the development version of flytekit.

File: _blogs/motherduck/ecommerce_wf.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    secret_requests=[Secret(group=None, key="md_token")],
+    container_image=image,
+)
+
+
+@task(container_image=image, enable_deck=True)
+def query_result_report(
+        sales_trends_result: pd.DataFrame,
+        elasticity_result: pd.DataFrame,
+        customer_segmentation_result: pd.DataFrame,
```
Comment (samhita-alla) [inline _blogs/motherduck/ecommerce_wf.py:66]: indentation

File: _blogs/motherduck/ecommerce_wf.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+from duckdb_artifacts import RecentEcommerceData, on_recent_ecommerce_data
+from openai_tools import get_tools, GPT_MODEL, DUCKDB_FUNCTION_NAME
+from plots import make_elasticity_plot, make_sales_trend_plot, create_transactions_plot, \
+    create_spend_plot
+from queries import sales_trends_query, elasticity_query, customer_segmentation_query
+
+image = ImageSpec(
+    registry=os.environ.get("DOCKER_REGISTRY", None),
+    packages=["union==0.1.68", "pandas==2.2.2", "plotly==5.23.0", "flytekitplugins-openai", "flytekitplugins-duckdb"],
```
Comment (samhita-alla) [inline _blogs/motherduck/ecommerce_wf.py:23]: are you even using flytekitplugins-openai plugin?

File: _blogs/motherduck/ecommerce_wf.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        return response_message.content, "No query."
+
+
+@workflow
+def wf(recent_data: pd.DataFrame = RecentEcommerceData.query(), prompt: Optional[str] = None) -> str:
+    # Answer prompt
+    answer, query = check_prompt(recent_data=recent_data, prompt=prompt)
+
+    # Make plot
+    sales_trends_result = sales_trends_query_task(mydf=recent_data)
```
Comment (samhita-alla) [inline _blogs/motherduck/ecommerce_wf.py:196]: should this be a separate workflow as it seems unrelated to the `check_prompt` dynamic workflow?

File: _blogs/motherduck/openai_tools.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        # Format the schema for this table
+        schema_info[table_name] = [{"column_name": col[0], "column_type": col[1]} for col in schema]
+    # Convert the schema info to JSON format
+    schema_json = json.dumps(schema_info, indent=4)
+    return schema_json
+
+def get_tools(con: duckdb.DuckDBPyConnection):
+    database_schema_string = get_schema(con)
+
+    tools = [
```
Comment (samhita-alla) [inline _blogs/motherduck/openai_tools.py:46]: can we have one more tool here? i think it'd be useful to show openai model's selection between two queries. seems reasonable?

File: _blogs/motherduck/ecommerce_wf.py (hunk 3 above)
Comment (dansola) [reply _blogs/motherduck/ecommerce_wf.py:23]: ah no I'm not

File: _blogs/motherduck/openai_tools.py (hunk 5 above)
Comment (dansola) [reply _blogs/motherduck/openai_tools.py:46]: Hmm right now it's letting chatgpt create the query rather than select the query. If it doesn't think the prompt is relevant to the description of the tool, it does not select it and it just answers the prompt like you were asking a regular chatgpt agent. should i clarify this in the blogpost?

This is why added runtime queries to the duckdb plugin btw - so we can have chatgpt actually write the query.

Comment (samhita-alla) [reply _blogs/motherduck/openai_tools.py:46]: i meant let's have two tools for two separate contexts. i understand that the tool generates duckdb query. for us to show the benefits of function calling, we must have a minimum of two tools. how about we add another tool that queries a different table?
